src/sources/udp.rs

```diff
@@ -0,0 +1,220 @@
+use serde::{Deserialize, Serialize};
+use std::{io, net::SocketAddr};
+use string_cache::DefaultAtom as Atom;
+use tokio::{
+    self,
+    codec::{BytesCodec, Decoder},
+    net::{UdpFramed, UdpSocket},
+};
+use tracing::field;
+
+/// UDP processes messages per packet, where messages are separated by newline.
+#[derive(Deserialize, Serialize, Debug, Clone)]
+#[serde(deny_unknown_fields)]
+pub struct UdpConfig {
+    pub address: SocketAddr,
+    pub host_key: Option<Atom>,
+}
+
+impl UdpConfig {
+    pub fn new(address: SocketAddr) -> Self {
+        Self {
+            address,
+            host_key: None,
+        }
+    }
+}
+
+#[typetag::serde(name = "udp")]
+impl SourceConfig for UdpConfig {
+    fn build(
+        &self,
+        _name: &str,
+        _globals: &GlobalOptions,
+        out: mpsc::Sender<Event>,
+    ) -> Result<super::Source, String> {
+        let host_key = self.host_key.clone().unwrap_or(event::HOST.clone());
+        Ok(udp(self.address, host_key, out))
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+}
+
+pub fn udp(address: SocketAddr, host_key: Atom, out: mpsc::Sender<Event>) -> super::Source {
+    let out = out.sink_map_err(|e| error!("error sending event: {:?}", e));
+
+    Box::new(
+        future::lazy(move || {
+            let socket = UdpSocket::bind(&address).expect("failed to bind to udp listener socket");
+
+            info!(message = "listening.", addr = &field::display(address),);
+
+            future::ok(socket)
+        })
+        .and_then(move |socket| {
+            let host_key = host_key.clone();
+            let mut new_line_decoder = BytesDelimitedCodec::new(b'\n');
+            UdpFramed::new(socket, BytesCodec::new())
+                .map(move |(mut bytes, addr)| {
+                    // A packet has come.
+                    // UDP processes messages per packet, where messages are separated by newline.
+                    let mut events = Vec::new();
+                    while let Some(Some(line)) = new_line_decoder
+                        .decode_eof(&mut bytes)
+                        .map_err(|e| error!("{}", e))
+                        .ok()
+                    {
+                        // One message
+                        let mut event = Event::from(line);
+
+                        event
+                            .as_mut_log()
+                            .insert_implicit(host_key.clone().into(), addr.to_string().into());
+
+                        trace!(
+                            message = "Received one event.",
+                            event = field::debug(&event)
+                        );
+
+                        events.push(event);
+                    }
+                    iter_ok(events)
+                })
+                // Flatten messages from single packet
+                .flatten()
+                .map_err(|e: io::Error| error!("error reading datagram: {:?}", e))
```

Comment:
```suggestion
                .map_err(|error: io::Error| error!(message = "error reading datagra.", %error))
```

Reply:
m in datagram is missing, so I will add this separately.